### ISSUE TITLE
Setting ob_tran CRS

### ISSUE THREAD

**Issue by adrfantini** [r-spatial/stars#52]
I have a `netcdf` file ([link](https://drive.google.com/open?id=1RQMv5wDWv4_8Pq_2zuJvqhbKRT9_KYjr)) of which I know the proper CRS, but the CRS is not encoded in the file itself:
`+proj=ob_tran +o_proj=longlat +o_lat_p=29.0 +o_lon_p=0.0 +lon_0=15.0 +a=6367470 +b=6367470`

`stars`, however, refuses to use it:
```
st_crs(s) = '+proj=ob_tran +o_proj=longlat +o_lat_p=29.0 +o_lon_p=0.0 +lon_0=15.0 +a=6367470 +b=6367470'
GDAL cannot import PROJ.4 string `+proj=ob_tran +o_proj=longlat +o_lat_p=29.0 +o_lon_p=0.0 +lon_0=15.0 +a=6367470 +b=6367470': returning missing CRS
```

This is very similar to what happened in  r-spatial/sf#651 and has to do, as far as I understand it, with limited support in GDAL for the transformation `ob_tran`.

In the case of `sf`, a workaround using `lwgeom` was possible. Is there anything similar that can be done for `stars`?

**Comment by edzer** [comment]
Currently, we can read this file as a curvilinear grid, but when read it as a regular GDAL file with `read_stars` there is no information about the georeference, at all.

**Comment by edzer** [comment]
`stars` is not alone here:
```r
> r = readGDAL("hmr_big.nc")
hmr_big.nc has GDAL driver netCDF 
and has 1361 rows and 1286 columns
Warning message:
In readGDAL("hmr_big.nc") : GeoTransform values not available
```
so in order to test this, please provide a netcdf file that at least gives a longlat grid, not just a matrix, when read through GDAL. Or give the key to this (offset, delta for x and y). As far as I understand, we shouldn't read these from the long and lat matrices, because they would contain the non-obtran transformed coordinates (ie the curvilinear grid).

**Comment by adrfantini** [comment]
Ok, I'll test this out and find a proper example in a few days.

**Comment by adrfantini** [comment]
[Here](https://drive.google.com/file/d/1wINjQkEDNX-uZUfX6vpppX-k9caPJob3/view?usp=sharing) you can find a time slice of an example netCDF file from one EURO-CORDEX model. The grid is a rotated one, defined as:

```
        int rotated_latitude_longitude ;
                rotated_latitude_longitude:grid_mapping_name = "rotated_latitude_longitude" ;
                rotated_latitude_longitude:grid_north_pole_latitude = 39.25 ;
                rotated_latitude_longitude:grid_north_pole_longitude = -162. ;
                rotated_latitude_longitude:north_pole_grid_longitude = 0. ;
```
`lat` and `lon` arrays are included. The CRS is not indicated explicitly. GDAL cannot understand the SRS of the file.

Can this help?

**Comment by edzer** [comment]
What would be the proj4string for these datasets?

**Comment by adrfantini** [comment]
In theory it should be the same as previously discussed in r-spatial/sf#651, so something along the lines of `"+proj=ob_tran +o_proj=longlat +o_lon_p=-162 +o_lat_p=39.25 +lon_0=175 +to_meter=0.01745329"`

**Comment by edzer** [comment]
This uses the forward projection trick also used in the issue you refer to. 
```r
library(stars)
# Loading required package: abind
# Loading required package: sf
# Linking to GEOS 3.6.2, GDAL 2.2.3, PROJ 4.9.3
(r = read_stars("example_ob_tran.nc"))
# stars object with 3 dimensions and 1 attribute
# attribute(s):
#  example_ob_tran.nc 
#  Min.   :247.0      
#  1st Qu.:267.4      
#  Median :275.9      
#  Mean   :274.5      
#  3rd Qu.:282.6      
#  Max.   :292.6      
# dimension(s):
#      from  to              offset delta  refsys point values    
# x       1 106              -28.43  0.44      NA    NA   NULL [x]
# y       1 103               21.89 -0.44      NA    NA   NULL [y]
# time    1   1 1979-01-16 12:00:00    NA POSIXct    NA   NULL    
```
First the path using the `lon` and `lat` subdatasets to directly create a curvilinear grid:
```r
lon = read_stars('NETCDF:"example_ob_tran2.nc":lon')
lat = read_stars('NETCDF:"example_ob_tran2.nc":lat')
rr = st_as_stars(r, curvilinear = list(x = lon[[1]], y = lat[[1]]))
rr
# stars object with 3 dimensions and 1 attribute
# attribute(s):
#  example_ob_tran.nc 
#  Min.   :247.0      
#  1st Qu.:267.4      
#  Median :275.9      
#  Mean   :274.5      
#  3rd Qu.:282.6      
#  Max.   :292.6      
# dimension(s):
#      from  to              offset delta                       refsys point
# x       1 106                  NA    NA +proj=longlat +datum=WGS8...    NA
# y       1 103                  NA    NA +proj=longlat +datum=WGS8...    NA
# time    1   1 1979-01-16 12:00:00    NA                      POSIXct    NA
#                                       values    
# x    [106x103] -44.1407 [°], ..., 64.404 [°] [x]
# y    [106x103] 22.1994 [°], ..., 72.4199 [°] [y]
# time                                    NULL    
# curvilinear grid
plot(rr, axes = TRUE, border = NA)
```
![x1](https://user-images.githubusercontent.com/520851/47110191-72cc2c00-d250-11e8-8b6f-7cdc6123be56.png)
Next compute the curvilinear grid parametrically from the shifted grid (`r`), using the forward trick:
```r
obtran = "+proj=ob_tran +o_proj=longlat +o_lon_p=-162 +o_lat_p=39.25 +lon_0=180 +to_meter=0.01745329"
st_crs(r) = 4326
r.ll = st_transform(r, obtran)
st_crs(r.ll) = 4326
plot(r.ll, axes = TRUE, border = NA)
```
![x2](https://user-images.githubusercontent.com/520851/47110226-811a4800-d250-11e8-9f2b-4903ec8583dd.png)
```r
r.ll
# stars object with 3 dimensions and 1 attribute
# attribute(s):
#  example_ob_tran.nc 
#  Min.   :247.0      
#  1st Qu.:267.4      
#  Median :275.9      
#  Mean   :274.5      
#  3rd Qu.:282.6      
#  Max.   :292.6      
# dimension(s):
#      from  to              offset delta                       refsys point
# x       1 106                  NA    NA +proj=longlat +datum=WGS8...    NA
# y       1 103                  NA    NA +proj=longlat +datum=WGS8...    NA
# time    1   1 1979-01-16 12:00:00    NA                      POSIXct    NA
#                               values    
# x    [106x103] -44.1407, ..., 64.404 [x]
# y      [106x103] 22.1994, ..., 72.42 [y]
# time                            NULL    
# curvilinear grid

st_bbox(rr)
# Units: [°]
#      xmin      ymin      xmax      ymax 
# -44.14069  22.19937  64.40398  72.41995 
st_bbox(r.ll)
#      xmin      ymin      xmax      ymax 
# -44.14070  22.19937  64.40399  72.41996 
```

**Comment by adrfantini** [comment]
I guess that now after #65 the another approach would be:
`r2 = read_stars('example_ob_tran.nc', curvilinear=c('lon', 'lat'))`
Is one approach more 'correct' than the other?

**Comment by edzer** [comment]
the obtran PROJ string does not come from the file, `lon` and `lat` do. Using only stuff that comes from the file feels easier. Having said that, the obtran parameters again may be somewhere in the metadata. `lon` and `lat` arrays only work when they are present.

**Comment by adrfantini** [comment]
I am having issues with `curvilinear` for [this](https://drive.google.com/open?id=1ptt_FpGH1OJGZK80qUPevOYX4BKi2zJl) file with rotated grid.
```R
fn = 'historical.nc'
fn %>% read_stars(var = 'Q100', curvilinear=c('lon', 'lat'))
#time_bnds, trying to read file: netCDF:NETCDF:"/home/clima-archive4-b/afantini/regcm_simulations/EURO-CORDEX/change/italy_change/mrro/faster_way/data_hadgem/yearmax/historical.nc":time_bnds:lon
#Error in CPL_read_gdal(as.character(x), as.character(options), as.character(driver),  : 
  file not found

fn %>% read_ncdf(var = 'Q100', curvilinear=c('lon', 'lat'))
#Error: length(d) == length(dim(x[[1]])) is not TRUE

lon = fn %>% read_ncdf(var = 'lon', quiet = TRUE)
lat = fn %>% read_ncdf(var = 'lat', quiet = TRUE)
ll = setNames(c(hlon, hlat), c("x", "y"))
fn %>% read_ncdf(var = 'Q100', quiet = TRUE)  %>% st_as_stars(curvilinear = ll)
#Error: length(d) == length(dim(x[[1]])) is not TRUE
```
Can you reproduce this?

**Comment by edzer** [comment]
```r
fn = 'historical.nc'
lon = fn %>% read_ncdf(var = 'lon', quiet = TRUE)
lat = fn %>% read_ncdf(var = 'lat', quiet = TRUE)
lo = st_set_dimensions(lon, names = c("x", "y"))
la = st_set_dimensions(lat, names = c("x", "y"))
ll = setNames(c(lo, la), c("x", "y"))
x = fn %>% read_stars(sub = 'Q100', quiet = TRUE)  %>% st_as_stars(curvilinear = ll)
```
Any suggestions how we can make this easier?

And, any idea what went wrong?
![x](https://user-images.githubusercontent.com/520851/51550254-4c7be500-1e6c-11e9-8acf-55b8ac14aabc.png)

**Comment by adrfantini** [comment]
Shouldn't we be able to read this by just doing `fn %>% read_stars(sub = 'Q100', curvilinear = c('lon', 'lat'))`?

The inversion might be related to #73

**Comment by adrfantini** [comment]
If you read `lat` and `lon` with `read_stars` instead of `read_ncdf`, it plots fine. Alternatively, I assume you could read everything with `read_ncdf`, but I am getting a `Error: length(d) == length(dim(x[[1]])) is not TRUE` if trying to plot or display `x = fn %>% read_ncdf(var = 'Q100', quiet = TRUE)  %>% st_as_stars(curvilinear = ll)`

**Comment by edzer** [comment]
Thanks for sorting out! @mdsumner this may be related that in lon and lat _both_ x and y dimensions have a negative delta (cellsize)?

**Comment by mdsumner** [comment]
Is that the GDAL default? I remember it would always be upside down if the "default" transform was applied, where offsets both zero and both 1. It was the wrong orientation for a simple png image. I'll explore

**Comment by edzer** [comment]
First time I see a negative x offset! It is simply what the netcdf gdal driver communicates; I guess it comes from the ordering of the array in the netcdf file which is, afaict, not prescribed.

**Comment by edzer** [comment]
```r
fn = 'historical.nc'
fn %>% read_stars(sub = 'Q100', curvilinear = c('lon', 'lat')) %>% plot
```
now works.

**Comment by edzer** [comment]
What happens if you add `driver = NULL` to the `read_stars` command?

**Comment by adrfantini** [comment]
Same error.

**Comment by edzer** [comment]
This is after installing the branch with
```r
remotes::install_github("r-spatial/stars", "values")
```
right?

**Comment by adrfantini** [comment]
Dang! Sorry, I missed that. Yep, can confirm it's working if using that branch. I believe we can close this?
The `Error: length(d) == length(dim(x[[1]])) is not TRUE` with `read_ncdf` persists but maybe that's material for another issue?

**Comment by edzer** [comment]
Why don't we get the full time series in the stars object?

**Comment by edzer** [comment]
Ah, sorry, that only concerns the `mrro` variable.
![x](https://user-images.githubusercontent.com/520851/51619504-bbbb0d00-1f30-11e9-9f23-1a982eaa61c9.png)

**Comment by adrfantini** [comment]
Ah, see the damned 360-day calendar issue right there.

**Comment by edzer** [comment]
I noticed - time for its own temporal reference system. Did you mention there was an R package to convert back and forth between calendar time and 360-day/year model time?

**Comment by mdsumner** [comment]
@edzer where do you see a negative x? I see default geotransform on time, and typical +x/-y on the rest

```
# xoffset, xdelta, yskew, yres, xskew, ydelta (internal GDAL geotransform arrangement)
$time
[1] 0 1 0 0 0 1

$lon
[1] -9.51  0.11  0.00 -2.20  0.00 -0.11

$lat
[1] -9.51  0.11  0.00 -2.20  0.00 -0.11

$mrro
[1] -9.51  0.11  0.00 -2.20  0.00 -0.11

$Q100
[1] -9.51  0.11  0.00 -2.20  0.00 -0.11
```

in gdalinfo terms

```R
Driver: netCDF/Network Common Data Format
Files: historical.nc
Size is 94, 114
Coordinate System is `'
Origin = (-9.509999923167690,-2.200000113090583)
Pixel Size = (0.109999998923271,-0.110000002700671)
```

But, I'd assume to ignore the geotransform in curvilinear case anyway - right?.  Here using raster (divorced from how I tackle this in ocean model data - ignoring the extent/crs, but analogous) and using the quadmesh grid/base plot: 

```R
f <- "historical.nc"
r <- raster::raster(f, varname = "mrro")
coords <- raster::brick(raster::raster(f, varname = "lon"), 
                        raster::raster(f, varname = "lat"))
library(quadmesh)
mesh_plot(r, coords = coords)  
maps::map(add = TRUE)
```

![image](https://user-images.githubusercontent.com/4107631/51630727-05dad780-1f9f-11e9-96a4-a4abd8dcfdae.png)

**Comment by mdsumner** [comment]
Oh! Yes, so easy to mix up - worldfile order, GDAL order, raster extent order ...

Yes both raster/quadmesh are CRAN versions. 


Hmm, I'm seeing that error in the netcdf pathway - so I'll investigate that(!): 

```R
x <- read_ncdf(fn, curvilinear = c("lon", "lat"))
x
#Error in dim.stars(x) : length(d) == length(dim(x[[1]])) is not TRUE
```

**Comment by edzer** [comment]
First shot at PCICt support:
```r
library(stars)
# Loading required package: abind
# Loading required package: sf
# Linking to GEOS 3.7.0, GDAL 2.3.2, PROJ 5.2.0
fn = 'historical.nc'
(xx = read_stars(fn, sub = 'mrro', curvilinear = c("lon", "lat")))
# lon, 
# lat, 
# mrro, 
# stars object with 3 dimensions and 1 attribute
# attribute(s):
#  historical.nc":mrro [kg/m^2/s]
#  Min.   :0                     
#  1st Qu.:0                     
#  Median :0                     
#  Mean   :0                     
#  3rd Qu.:0                     
#  Max.   :0                     
#  NA's   :173400                
# dimension(s):
#      from  to     offset    delta                       refsys point
# x       1  94         NA       NA +proj=longlat +datum=WGS8...    NA
# y       1 114         NA       NA +proj=longlat +datum=WGS8...    NA
# time    1  30 1976-07-01 360 days                        PCICt    NA
#                                    values    
# x    [94x114] 3.91253 [°],...,19.1685 [°] [x]
# y     [94x114] 35.3469 [°],...,48.495 [°] [y]
# time                                 NULL    
# curvilinear grid
plot(xx[,,,1:12])
# Warning messages:
# 1: In seq(from = x$offset + (x$from - 1 + where) * x$delta, by = x$delta,  :
#   Incompatible methods ("+.PCICt", "Ops.difftime") for "+"
# 2: In seq(from = x$offset + (x$from - 1 + where) * x$delta, by = x$delta,  :
#   Incompatible methods ("+.PCICt", "Ops.difftime") for "+"
```
Look at the labels:
![x](https://user-images.githubusercontent.com/520851/51640403-a6aba180-1f63-11e9-82bc-a354a8d59cab.png)

**Comment by adrfantini** [comment]
Wonderful! Should probably continue this in #29
